design/ceph/nvmeof-recoverer.md

```diff
@@ -0,0 +1,184 @@
+
+## Proposal Details
+
+### Overview
+
+We propose a design, called NVMe-oF recoverer. NVMe-oF is an emerging storage network where the storage devices can be attached to any storage node using RDMA, TCP or Fiber channel. In this environment, the failure caused by a server node can be recoverable because the OSD can be re-spawned to another server node with the storage device attached to the failed server node through NVMe-oF.
+
+To address these challenges, we recommend a phased implementation consisting of four key steps:
+
+1. **Implement an Explicit Command for Manual OSD Relocation**
+2. **Automate OSD Relocation Upon Failure**
+3. **Modify CRUSH Map to Minimize Data Rebalancing**
+4. **Improve OSD Scheduling Based on Node Workloads**
+
+Each step can be implemented and contributed individually, allowing for incremental progress and testing.
+
+### Step 1: Implement an Explicit Command for Manual OSD Relocation
+
+#### Description
+
+Introduce explicit commands or scripts that allow administrators to manually deploy and relocate OSD pods using NVMe-oF devices. These tools will streamline the management of NVMe-oF devices and OSD pods in standard scenarios and during failures.
+
+#### Implementation Steps
+
+- **Develop a Script or Command-line Tool**:
+  - **For Connecting NVMe-oF Devices and Deploying OSD Pods**:
+    - Retrieve NVMe-oF target information from predefined sources (e.g., configuration files, environment variables).
+      - Key parameters: IP, Port, Subsystem NQN (SubNQN), and attachable nodes.
+    - Connect devices to target nodes using `nvme-cli` or appropriate tools.
+    - Update the `CephCluster` Custom Resource Definition (CRD) with fabric device details to facilitate automated OSD deployment.
+  - **For Handling OSD Relocation in Failure Scenarios**:
+    - Remove the faulted OSD pod.
+    - Disconnect the device from the failed node and reconnect it to a healthy node.
+    - Update the `CephCluster` CRD to reflect the new device-node mapping.
+- **Usage Example**:
+
+  ```bash
+  kubectl exec -it rook-ceph-operator-xxxxx -- osd_move node-failed node-healthy
+  ```
+
+#### Benefits
+
+- Provides immediate control for administrators.
+- Simplifies the process of OSD deployment and relocation.
+
+### Step 2: Automate OSD Relocation Upon Failure
+
+#### Description
+
+Enhance the Rook operator to automatically detect OSD pod failures and relocate the affected OSDs to healthy nodes, reducing manual intervention and downtime.
+
+#### Implementation Steps
+
+- **Discovery of Connected Fabric Devices**:
+  - **Scan for Previously Connected Devices**:
+    - During initialization, use Kubernetes Jobs to scan all nodes for previously connected NVMe-oF devices.
+    - Perform device discovery using `nvme-cli`.
+  - **Identify Devices by SubNQN**:
+    - Use SubNQN to verify the connection status of devices and ensure correct node mapping.
+- **Detect OSD Pod Failures**:
```

Comment:
What type of pod failures? If the node goes down, K8s can automatically reschedule the pod to another node.